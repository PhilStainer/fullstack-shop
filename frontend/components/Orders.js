import Link from 'next/link'
import {useQuery} from '@apollo/react-hooks'
import {formatDistance} from 'date-fns'

import ORDERS_QUERY from '#root/graphql/orders.query'

import StyledOrders from '#root/components/styles/StyledOrders'
import formatPrice from '#root/utils/formatPrice'

const Orders = () => {
  const {data, loading, error} = useQuery(ORDERS_QUERY)

  if (loading) return <p>Loading orders...</p>
  if (error) return <p>{error.message}</p>

  return (
    <StyledOrders>
      <h2>You have {data.orders.length} orders</h2>

      <ul>
        <li>
          <a>
            <p>Ordered</p>
            <p>Order Number</p>
            <p>Items</p>
            <p>Products</p>
            <p>Total</p>
          </a>
        </li>

        {data.orders.map(order => (
          <li key={order._id} title="View order">
            <Link href={{pathname: '/order', query: {id: order._id}}}>
              <a>
                <p>
                  {formatDistance(new Date(order.createdAt), new Date(), {
                    addSuffix: true,
                  })}
                </p>
                <p>
                  <strong>{order._id}</strong>
                </p>
                <p>{order.items.reduce((a, b) => a + b.quantity, 0)}</p>
                <p>{order.items.length}</p>
                <p>{formatPrice(order.total)}</p>
              </a>
            </Link>
          </li>
        ))}
      </ul>
    </StyledOrders>
  )
}

export default Orders
