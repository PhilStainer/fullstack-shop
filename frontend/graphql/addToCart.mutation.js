import gql from 'graphql-tag'

const ADD_TO_CART_MUTATION = gql`
  mutation($id: ID!) {
    addToCart(id: $id) {
      _id
      quantity
      item {
        _id
        title
        description
        imageUrl
        price
      }
    }
  }
`

export default ADD_TO_CART_MUTATION
