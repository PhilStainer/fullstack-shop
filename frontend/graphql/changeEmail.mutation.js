import gql from 'graphql-tag'

const CHANGE_PASSWORD_MUTATION = gql`
  mutation($password: String!, $email: String!, $confirmEmail: String!) {
    changeEmail(
      password: $password
      email: $email
      confirmEmail: $confirmEmail
    ) {
      status
      message
    }
  }
`

export default CHANGE_PASSWORD_MUTATION
