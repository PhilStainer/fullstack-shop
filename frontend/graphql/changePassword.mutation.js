import gql from 'graphql-tag'

const CHANGE_PASSWORD_MUTATION = gql`
  mutation(
    $currentPassword: String!
    $password: String!
    $confirmPassword: String!
  ) {
    changePassword(
      currentPassword: $currentPassword
      password: $password
      confirmPassword: $confirmPassword
    ) {
      status
      message
    }
  }
`

export default CHANGE_PASSWORD_MUTATION
