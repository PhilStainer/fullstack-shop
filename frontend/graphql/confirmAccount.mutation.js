import gql from 'graphql-tag'

const CONFIRM_ACCOUNT_MUTATION = gql`
  mutation($confirmToken: String!) {
    confirmAccount(confirmToken: $confirmToken) {
      status
      message
    }
  }
`

export default CONFIRM_ACCOUNT_MUTATION
