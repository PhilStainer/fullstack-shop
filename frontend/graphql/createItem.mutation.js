import gql from 'graphql-tag'

const CREATE_ITEM_MUTATION = gql`
  mutation(
    $title: String!
    $description: String!
    $imageUrl: String!
    $price: Int!
  ) {
    createItem(
      title: $title
      description: $description
      imageUrl: $imageUrl
      price: $price
    ) {
      _id
    }
  }
`

export default CREATE_ITEM_MUTATION
