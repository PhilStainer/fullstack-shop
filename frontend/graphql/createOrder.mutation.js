import gql from 'graphql-tag'

const CREATE_ORDER_MUTATION = gql`
  mutation($token: String!) {
    createOrder(token: $token) {
      _id
    }
  }
`

export default CREATE_ORDER_MUTATION
