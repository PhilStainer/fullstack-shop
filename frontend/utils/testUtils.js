export const fakeUser = () => ({
  _id: '12345',
  name: 'Test Name',
  email: 'test@email.com',
})
