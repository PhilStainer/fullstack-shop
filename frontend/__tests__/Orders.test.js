import {render, waitFor} from '@testing-library/react'
import {MockedProvider} from '@apollo/react-testing'
import {GraphQLError} from 'graphql'

import ORDERS_QUERY from '#root/graphql/orders.query'

import Orders from '#root/components/Orders'
import {fakeOrder} from '#root/utils/testUtils'

const query = {id: 'order id'}

test('fetch items on mount', async () => {
  const {getByText} = render(
    <MockedProvider>
      <Orders />
    </MockedProvider>,
  )

  await waitFor(() => {
    expect(getByText(/loading orders/i)).toBeInTheDocument()
  })
})

test('renders error on graphql error', async () => {
  const mock = {
    request: {query: ORDERS_QUERY},
    result: jest.fn(() => ({
      errors: [new GraphQLError('Error finding orders...')],
    })),
  }

  const {getByText} = render(
    <MockedProvider mocks={[mock]} addTypename={false}>
      <Orders />
    </MockedProvider>,
  )

  await waitFor(() => {
    expect(getByText(/error finding orders/i)).toBeInTheDocument()
  })
})

test('renders orders', async () => {
  const mock = {
    request: {query: ORDERS_QUERY},
    result: jest.fn(() => ({data: {orders: [fakeOrder()]}})),
  }

  const {container} = render(
    <MockedProvider mocks={[mock]} addTypename={false}>
      <Orders />
    </MockedProvider>,
  )

  await waitFor(() => {
    expect(container).toMatchSnapshot()
  })
})
