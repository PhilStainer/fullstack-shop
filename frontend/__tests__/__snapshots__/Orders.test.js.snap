// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders orders 1`] = `
<div>
  <p>
    Loading orders...
  </p>
</div>
`;

exports[`renders orders 2`] = `
<div>
  <div
    class="StyledOrders-sc-90v5kd-0 CWRYh"
  >
    <h2>
      You have 
      1
       orders
    </h2>
    <ul>
      <li>
        <a>
          <p>
            Ordered
          </p>
          <p>
            Order Number
          </p>
          <p>
            Items
          </p>
          <p>
            Products
          </p>
          <p>
            Total
          </p>
        </a>
      </li>
      <li
        title="View order"
      >
        <a
          href="/order?id=12345"
        >
          <p>
            about 2 years ago
          </p>
          <p>
            <strong>
              12345
            </strong>
          </p>
          <p>
            1
          </p>
          <p>
            1
          </p>
          <p>
            Â£50
          </p>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
